reaper/reaper.go

```diff
@@ -17,141 +17,153 @@
-				fatalError("Cannot read GET /v2/service_plans/[service plan GUID]/service_instances response body: %s", err, fatal)
-			}
+		output <- services[0]
+	}()
 
-			var servicePlanListServiceInstancesResp ServicePlanListServiceInstancesResp
-			err = json.Unmarshal(body, &servicePlanListServiceInstancesResp)
-			if err != nil {
-				fatalError("Invalid GET /v2/service_plans/[service plan GUID]/service_instances response JSON", err, fatal)
-			}
+	return output
+}
+
+func (r *reaper) freeServicePlansOf(services <-chan cloudfoundry.Service) <-chan cloudfoundry.ServicePlan {
+	output := make(chan cloudfoundry.ServicePlan)
+
+	go func() {
+		defer close(output)
+
+		for service := range services {
+			servicePlans, servicePlanErrors := r.cf.GetServicePlans(service.Metadata.Guid, MaximumResultsPerPage)
 
-			for _, si := range servicePlanListServiceInstancesResp.Resources {
-				if expired(si.Metadata.CreatedAt, expiryInterval, fatal) {
-					if reap {
-						// TODO: delete the service instance and add force option to do recursive delete
-					} else {
-						fmt.Printf("%s %s", si.Entity.Name, si.Metadata.Guid)
-					}
+			for servicePlan := range servicePlans {
+				if servicePlan.Entity.Free {
+					output <- servicePlan
 				}
 			}
 
+			mergeErrors(servicePlanErrors, r.errorChannel)
 		}
-	}
-}
+	}()
 
-func fatalError(message string, err error, fatal func(string)) {
-	fatal(fmt.Sprintf("%s: %s", message, err))
+	return output
 }
 
-func expired(creationTimeString string, expiryInterval time.Duration, fatal func(string)) bool {
-	creationTime, err := time.Parse(time.RFC3339, creationTimeString)
-	if err != nil {
-		fatalError("Invalid service instance creation time", err, fatal)
-	}
-	expiryTime := creationTime.Add(expiryInterval)
-	return time.Now().After(expiryTime)
-}
+func (r *reaper) expiredInstancesOf(servicePlans <-chan cloudfoundry.ServicePlan) <-chan cloudfoundry.ServiceInstance {
+	output := make(chan cloudfoundry.ServiceInstance)
 
-type Metadata struct {
-	Guid      string
-	CreatedAt string `json:"created_at"`
-}
+	go func() {
+		defer close(output)
 
-type ListServicesResp struct {
-	Resources []struct {
-		Metadata Metadata
-	}
+		for servicePlan := range servicePlans {
+			serviceInstances, serviceInstanceErrors := r.cf.GetServicePlanInstances(servicePlan.Metadata.Guid, MaximumResultsPerPage)
+
+			for serviceInstance := range serviceInstances {
+				serviceInstanceExpired, err := expired(serviceInstance.Metadata.CreatedAt, r.expiryInterval, r.currentTime)
+				if err != nil {
+					r.errorChannel <- err
+					return
+				}
+
+				if serviceInstanceExpired {
+					output <- serviceInstance
+				}
+			}
+
+			mergeErrors(serviceInstanceErrors, r.errorChannel)
+		}
+	}()
+
+	return output
 }
 
-type ListServicePlansResp struct {
-	Resources []struct {
-		Metadata Metadata
-		Entity struct {
-			Name string
-			Free bool
+func (r *reaper) delete(serviceInstances <-chan cloudfoundry.ServiceInstance) {
+	go func() {
```

Comment:
Same as previous, as this is a infrequent batch operation, we might want to consider synchronous ops